### PR TITLE
#70 - First implementation of the support for multiple ASDs per Disk

### PR DESCRIPTION
**Pull request by khenderick**
Work in progress

### CODE REVIEW

File: source/app/api.py [hunk 1]
```diff
@@ -225,26 +193,214 @@ def delete_disk(disk):
...
+            if disk in mountpoints:
+                mountpoint = mountpoints[disk]
+                asds = API._list_asds_disk(disk, mountpoint)
+                for asd_id in asds:
+                    service_name = '{0}{1}'.format(API.ASD_SERVICE_PREFIX, asd_id)
+                    if ServiceManager.has_service(service_name, local_client):
+                        ServiceManager.stop_service(service_name, local_client)
+            Disks.remount_disk(disk)
+            mountpoints = FSTab.read()
+            if disk in mountpoints:
```
Comment (kvanhijf) [inline source/app/api.py]: Better to raise ```if disk not in mountpoints: raise```

File: source/app/api.py [hunk 2]
```diff
@@ -225,26 +193,214 @@ def delete_disk(disk):
...
+                                                     'detail': 'servicefailure'}
+                    API._asd_hateoas(asds[asd_id], disk, asd_id)
+        except OSError as ex:
+            API._log('Error collecting ASD information: {0}'.format(str(ex)))
+        return asds
+
+    @staticmethod
+    @get('/asds')
+    def list_asds():
+        asds = {}
```
Comment (kvanhijf) [inline source/app/api.py]: docstring plz

File: source/app/api.py [hunk 3]
```diff
@@ -225,26 +193,214 @@ def delete_disk(disk):
...
+            asd_id = ''.join(random.choice(string.ascii_letters + string.digits) for _ in range(32))
+            all_asds = {}
+            mountpoints = FSTab.read()
+            for _disk, mountpoint in mountpoints.iteritems():
+                all_asds.update(API._list_asds_disk(_disk, mountpoint))
+            mountpoint = mountpoints[disk]
+
+            # Prepare & start service
+            API._log('Setting up service for disk {0}'.format(disk))
+            homedir = '{0}/{1}'.format(mountpoint, asd_id)
```
Comment (kvanhijf) [inline source/app/api.py]: Directory exists at this point?

File: source/app/api.py [hunk 4]
```diff
@@ -225,26 +193,214 @@ def delete_disk(disk):
...
-            return {'_link': '/disks/{0}'.format(disk)}
+    @staticmethod
+    @get('/disks/<disk>/asds/<asd_id>')
+    def get_asd(disk, asd_id):
+        if EtcdConfiguration.exists(API.ASD_CONFIG.format(asd_id)):
+            asd = EtcdConfiguration.get(API.ASD_CONFIG.format(asd_id))
+            service_name = '{0}{1}'.format(API.ASD_SERVICE_PREFIX, asd_id)
+            if ServiceManager.has_service(service_name, local_client):
+                service_state = ServiceManager.get_service_status(service_name, local_client)
+                if service_state is False:
```
Comment (kvanhijf) [inline source/app/api.py]: Duplicate code block

File: source/app/api.py [hunk 5]
```diff
@@ -225,26 +193,214 @@ def delete_disk(disk):
...
+        if ServiceManager.has_service(service_name, local_client):
+            ServiceManager.restart_service(service_name, local_client)
+
+    @staticmethod
+    def _delete_asd(asd_id, mountpoint):
+        service_name = '{0}{1}'.format(API.ASD_SERVICE_PREFIX, asd_id)
+        if ServiceManager.has_service(service_name, local_client):
+            ServiceManager.stop_service(service_name, local_client)
+            ServiceManager.remove_service(service_name, local_client)
+        check_output('rm -rf {0}/{1}'.format(mountpoint, asd_id), shell=True)
```
Comment (kvanhijf) [inline source/app/api.py]: If asd_id could be an empty string, you might end up in a dangerous situation

File: source/tools/disks.py [hunk 6]
```diff
@@ -129,31 +141,46 @@ def prepare_disk(disk, asd_id):
...
         """
-        print 'Cleaning disk {0}/{1}'.format(disk, asd_id)
+        print 'Cleaning disk {0}'.format(disk)
+        mountpoints = FSTab.read()
+        mountpoint = mountpoints[disk]
         FSTab.remove('/dev/disk/by-id/{0}-part1'.format(disk))
-        check_output('umount /mnt/alba-asd/{0} || true'.format(asd_id), shell=True)
-        check_output('rm -rf /mnt/alba-asd/{0} || true'.format(asd_id), shell=True)
+        check_output('umount {0} || true'.format(mountpoint), shell=True)
+        check_output('rm -rf {0} || true'.format(mountpoint), shell=True)
```
Comment (kvanhijf) [inline source/tools/disks.py]: You cowboy

File: source/app/api.py [hunk 7]
```diff
@@ -18,294 +18,233 @@
...
+        with FileMutex('disk_{0}'.format(disk)):
+            API._log('Got lock for restarting disk {0}'.format(disk))
+            mountpoints = FSTab.read()
+            if disk in mountpoints:
+                mountpoint = mountpoints[disk]
+                asds = ASDController.list_asds(mountpoint)
+                for asd_id in asds:
+                    ASDController.stop_asd(asd_id)
+            DiskController.remount_disk(disk)
+            mountpoints = FSTab.read()
```
Comment (kvanhijf) [inline source/app/api.py]: Mountpoints won't have changed at this point, so obsolete

File: source/app/api.py [hunk 8]
```diff
@@ -18,294 +18,233 @@
...
+            raise BadRequest('Disk could not be added')
+        return all_disks[disk]
 
     @staticmethod
     @post('/disks/<disk>/delete')
     def delete_disk(disk):
         """
         Delete a disk
-        :param disk: Guid of the disk
+        :param disk: Identifier of the disk
```
Comment (kvanhijf) [inline source/app/api.py]: Add ```type``` please

File: source/app/api.py [hunk 9]
```diff
@@ -18,294 +18,233 @@
...
-        return data
+        return all_disks[disk]
 
     @staticmethod
     @post('/disks/<disk>/add')
     def add_disk(disk):
         """
         Add a disk
-        :param disk: Guid of the disk
+        :param disk: Identifier of the disk
```
Comment (kvanhijf) [inline source/app/api.py]: Add ```type``` please

File: source/app/api.py [hunk 10]
```diff
@@ -18,294 +18,233 @@
...
+        API._log('Listing disks')
+        return DiskController.list_disks()
 
     @staticmethod
     @get('/disks/<disk>')
     def index_disk(disk):
         """
         Retrieve information about a single disk
-        :param disk: Guid of the disk
+        :param disk: Identifier of the disk
```
Comment (kvanhijf) [inline source/app/api.py]: Add ```type``` please

File: source/app/api.py [hunk 11]
```diff
@@ -18,294 +18,233 @@
...
+                    ASDController.remove_asd(asd_id, mountpoint)
+            DiskController.clean_disk(disk)
 
     @staticmethod
     @post('/disks/<disk>/restart')
     def restart_disk(disk):
         """
         Restart a disk
-        :param disk: Guid of the disk
+        :param disk: Identifier of the disk
```
Comment (kvanhijf) [inline source/app/api.py]: Add ```type``` please

File: source/app/api.py [hunk 12]
```diff
@@ -18,294 +18,233 @@
...
+    def list_asds_disk(disk):
+        """
+        Lists all ASDs on a given disk
+        :param disk: Identifier of the disk
+        """
+        mountpoints = FSTab.read()
+        if disk not in mountpoints:
+            raise BadRequest('Disk {0} is not yet initialized'.format(disk))
+        mountpoint = mountpoints[disk]
+        asds = ASDController.list_asds(mountpoint)
```
Comment (kvanhijf) [inline source/app/api.py]: Can be written in 1 line
```return ASDController.list_asds(mountpoint)```

File: source/app/api.py [hunk 13]
```diff
@@ -18,294 +18,233 @@
...
-                time.sleep(3)
-                if counter == max_counter:
-                    raise cpe
-            except Exception as ex:
-                raise ex
+    @post('/disks/<disk>/asds/<asd_id>/delete')
+    def asd_delete(disk, asd_id):
+        """
+        Deletes an ASD on a given Disk
+        :param disk: Idientifier of the Disk
```
Comment (kvanhijf) [inline source/app/api.py]: Idientifier Kenneth?

File: source/app/api.py [hunk 14]
```diff
@@ -18,294 +18,233 @@
...
-        data = API._list_disks()
-        data['_parent'] = '/'
-        data['_actions'] = []
-        return data
+        API._log('Listing disks')
+        return DiskController.list_disks()
 
     @staticmethod
     @get('/disks/<disk>')
     def index_disk(disk):
```
Comment (kvanhijf) [inline source/app/api.py]: Rename all ```disk``` to something more specific, eg: ```disk_id```

File: source/app/api.py [hunk 15]
```diff
@@ -18,294 +18,233 @@
...
+        :param disk: Idientifier of the Disk
+        :param asd_id: The ASD ID of the ASD to be removed
+        """
+        # Stop and remove service
+        API._log('Removing services for disk {0}'.format(disk))
+        mountpoints = FSTab.read()
+        if disk not in mountpoints:
+            raise BadRequest('Disk {0} is not yet initialized'.format(disk))
+        all_asds = {}
+        mountpoints = FSTab.read()
```
Comment (kvanhijf) [inline source/app/api.py]: Obsolete

File: source/controllers/asd.py [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+import json
+import math
+import random
+import signal
+import string
+import datetime
+from subprocess import check_output
+from source.tools.fstab import FSTab
+from source.tools.services.service import ServiceManager
+from source.tools.configuration import EtcdConfiguration
```
Comment (kvanhijf) [inline source/controllers/asd.py]: Honor our import sorting

File: source/controllers/asd.py [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+from source.tools.localclient import LocalClient
+
+local_client = LocalClient()
```
Comment (kvanhijf) [inline source/controllers/asd.py]: Don't instantiate variables outside classes

File: source/controllers/asd.py [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+
+
+class ASDController(object):
+    NODE_ID = os.environ['ASD_NODE_ID']
+    ASD_SERVICE_PREFIX = 'alba-asd-'
```
Comment (kvanhijf) [inline source/controllers/asd.py]: Consistency in namings: ```'alba-asd-{0}'```

File: source/controllers/asd.py [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+        asds = {}
+        try:
+            for asd_id in os.listdir(mountpoint):
+                if os.path.isdir('/'.join([mountpoint, asd_id])) and EtcdConfiguration.exists(ASDController.ASD_CONFIG.format(asd_id)):
+                    asds[asd_id] = EtcdConfiguration.get(ASDController.ASD_CONFIG.format(asd_id))
+                    output = check_output('ls {0}/{1}/ 2>&1 || true'.format(mountpoint, asd_id), shell=True)
+                    if 'Input/output error' in output:
+                        asds[asd_id]['state'] = {'state': 'error',
+                                                 'detail': 'ioerror'}
+                    else:
```
Comment (kvanhijf) [inline source/controllers/asd.py]: Less ```else``` stacks for more readability
if ```False```, then continue to diminish the indents

File: source/controllers/asd.py [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+                        service_name = '{0}{1}'.format(ASDController.ASD_SERVICE_PREFIX, asd_id)
+                        if ServiceManager.has_service(service_name, local_client):
+                            service_state = ServiceManager.get_service_status(service_name, local_client)
+                            if service_state is False:
+                                asds[asd_id]['state'] = {'state': 'error',
```
Comment (kvanhijf) [inline source/controllers/asd.py]: A dictionary with ```state``` and ```statedetail``` would be better imo without the outer ```state``` being required

File: source/controllers/asd.py [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+                                                     'detail': 'servicefailure'}
+        except OSError as ex:
+            ASDController._log('Error collecting ASD information: {0}'.format(str(ex)))
+        return asds
+
+    @staticmethod
+    def create_asd(disk):
+        all_asds = {}
+        mountpoints = FSTab.read()
+        for _disk, mountpoint in mountpoints.iteritems():
```
Comment (kvanhijf) [inline source/controllers/asd.py]: Can be replaced with ```itervalues```

File: source/controllers/asd.py [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+            all_asds.update(ASDController.list_asds(mountpoint))
+        mountpoint = mountpoints[disk]
+
+        # Fetch disk information
+        df_info = check_output('df -k {0} || true'.format(mountpoint), shell=True).strip()
```
Comment (kvanhijf) [inline source/controllers/asd.py]: Doesn't ```df``` has an option to only print the required field and also in the correct sizing?

File: source/controllers/asd.py [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+        asd_config = {'home': homedir,
+                      'node_id': ASDController.NODE_ID,
+                      'asd_id': asd_id,
+                      'capacity': asd_size,
+                      'log_level': 'info',
+                      'port': port}
+
+        if EtcdConfiguration.exists('{0}/extra'.format(ASDController.CONFIG_ROOT)):
+            data = EtcdConfiguration.get('{0}/extra'.format(ASDController.CONFIG_ROOT))
+            for extrakey in data:
```
Comment (kvanhijf) [inline source/controllers/asd.py]: Can be replaced by ```asd_config.update(data)```

File: source/controllers/disk.py [hunk 24]
```diff
@@ -129,39 +149,54 @@ def prepare_disk(disk, asd_id):
...
-        Remove the disk as ALBA device
+        Remove the disk
         :param disk: Disk ID
-        :param asd_id: ASD ID
         :return: None
         """
-        print 'Cleaning disk {0}/{1}'.format(disk, asd_id)
+        DiskController._log('Cleaning disk {0}'.format(disk))
+        mountpoints = FSTab.read()
+        mountpoint = mountpoints[disk]
```
Comment (kvanhijf) [inline source/controllers/disk.py]: Possible ```KeyError```

File: source/controllers/disk.py [hunk 25]
```diff
@@ -129,39 +149,54 @@ def prepare_disk(disk, asd_id):
...
         FSTab.remove('/dev/disk/by-id/{0}-part1'.format(disk))
-        check_output('umount /mnt/alba-asd/{0} || true'.format(asd_id), shell=True)
-        check_output('rm -rf /mnt/alba-asd/{0} || true'.format(asd_id), shell=True)
+        check_output('umount {0} || true'.format(mountpoint), shell=True)
+        check_output('rm -rf {0} || true'.format(mountpoint), shell=True)
```
Comment (kvanhijf) [inline source/controllers/disk.py]: You cowboy (2)

File: source/controllers/maintenance.py [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This module contains the maintenance controller (maintenance service logic)
+"""
+import json
+from source.tools.configuration import EtcdConfiguration
+from source.tools.services.service import ServiceManager
+from source.tools.localclient import LocalClient
```
Comment (kvanhijf) [inline source/controllers/maintenance.py]: Import order!!!

File: source/controllers/maintenance.py [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+
+local_client = LocalClient()
```
Comment (kvanhijf) [inline source/controllers/maintenance.py]: Don't instantiate variable outside classes

File: source/controllers/maintenance.py [hunk 28]
```diff
@@ -0,0 +1,73 @@
...
+
+
+class MaintenanceController(object):
+    MAINTENANCE_PREFIX = 'ovs-alba-maintenance'
```
Comment (kvanhijf) [inline source/controllers/maintenance.py]: Templates have been updated to ```alba-maintenance``` so this should be changed too.

File: source/controllers/maintenance.py [hunk 29]
```diff
@@ -0,0 +1,73 @@
...
+
+    @staticmethod
+    def remove_maintenance_service(name):
+        """
+        Remove a maintenance service with a specific name
+        :param name: Name of the service
+        """
+        if ServiceManager.has_service(name, local_client):
+            ServiceManager.stop_service(name, local_client)
+        ServiceManager.remove_service(name, local_client)
```
Comment (kvanhijf) [inline source/controllers/maintenance.py]: Will throw error if ```has_service``` returns ```False```

File: source/controllers/update.py [hunk 30]
```diff
@@ -0,0 +1,126 @@
...
+"""
+This module contains logic related to updates
+"""
+import os
+import time
+import datetime
+from subprocess import CalledProcessError
+from source.tools.packages.package import PackageManager
+from source.tools.services.service import ServiceManager
+from source.tools.localclient import LocalClient
```
Comment (kvanhijf) [inline source/controllers/update.py]: Respect import sorting
Don't instantiate variables outside classes

File: source/controllers/update.py [hunk 31]
```diff
@@ -0,0 +1,126 @@
...
+
+local_client = LocalClient()
+
+
+class UpdateController(object):
+    PACKAGE_NAME = 'openvstorage-sdm'
+    ASD_SERVICE_PREFIX = 'alba-asd-'
```
Comment (kvanhijf) [inline source/controllers/update.py]: Consistency @khenderick

File: source/controllers/update.py [hunk 32]
```diff
@@ -0,0 +1,126 @@
...
+        while True and counter < max_counter:
+            counter += 1
+            try:
+                PackageManager.update(local_client)
+                break
+            except CalledProcessError as cpe:
+                time.sleep(3)
+                if counter == max_counter:
+                    raise cpe
+            except Exception as ex:
```
Comment (kvanhijf) [inline source/controllers/update.py]: Not required

File: source/tools/services/fleetctl.py [hunk 33]
```diff
@@ -218,15 +241,21 @@ def _get_unit(fleet_name):
...
             if ae.code == 404:
                 # make error more descriptive
                 raise ValueError('Unit with name {0} not found'.format(fleet_name))
             raise RuntimeError('Fleet API error {0}'.format(ae))
 
     @staticmethod
     def _get_unit_name(name, client):
-        name = Systemd._get_name(name, client, '/opt/OpenvStorage/config/templates/systemd/')
+        try:
+            name = Systemd._get_name(name, client, '/opt/OpenvStorage/config/templates/systemd/')
```
Comment (kvanhijf) [inline source/tools/services/fleetctl.py]: Path should be changed

File: source/controllers/update.py (hunk 31 above)
Comment (khenderick) [inline source/controllers/update.py]: It's used by ```startswith```, so it's better this way.

File: source/controllers/asd.py [hunk 34]
```diff
@@ -0,0 +1,159 @@
...
+    @staticmethod
+    def create_asd(disk):
+        all_asds = {}
+        mountpoints = FSTab.read()
+        for mountpoint in mountpoints.values():
+            all_asds.update(ASDController.list_asds(mountpoint))
+        mountpoint = mountpoints[disk]
+
+        # Fetch disk information
+        disk_size = int(check_output('df -B 1 --output=size {0} || true'.format(mountpoint), shell=True).splitlines()[1])
```
Comment (kvanhijf) [inline source/controllers/asd.py]: Could give index out of range error in case ```|| True ``` is met

File: source/controllers/asd.py [hunk 35]
```diff
@@ -0,0 +1,159 @@
...
+    CONFIG_ROOT = '/ovs/alba/asdnodes/{0}/config'.format(NODE_ID)
+
+    _local_client = LocalClient()
+
+    @staticmethod
+    def _log(message):
+        print '{0} - {1}'.format(str(datetime.datetime.now()), message)
+
+    @staticmethod
+    def list_asds(mountpoint):
```
Comment (kvanhijf) [inline source/controllers/asd.py]: Missing docstrings

File: source/controllers/asd.py [hunk 36]
```diff
@@ -0,0 +1,159 @@
...
+                            asds[asd_id].update({'state': 'ok'})
+                    else:
+                        asds[asd_id].update({'state': 'error',
+                                             'state_detail': 'servicefailure'})
+        except OSError as ex:
+            ASDController._log('Error collecting ASD information: {0}'.format(str(ex)))
+        return asds
+
+    @staticmethod
+    def create_asd(disk):
```
Comment (kvanhijf) [inline source/controllers/asd.py]: Rename ```disk``` to ```disk_id```

File: source/controllers/maintenance.py [hunk 37]
```diff
@@ -0,0 +1,73 @@
...
+
+class MaintenanceController(object):
+    MAINTENANCE_PREFIX = 'alba-maintenance'
+    _local_client = LocalClient()
+
+    @staticmethod
+    def get_services():
+        """
+        Retrieve all configured maintenance service running on this node for each backend
+        :return: list
```
Comment (kvanhijf) [inline source/controllers/maintenance.py]: Returns a ```generator```

File: source/tools/services/systemd.py [hunk 38]
```diff
@@ -15,7 +15,7 @@
 """
 Systemd module
 """
-
+import datetime
```
Comment (kvanhijf) [inline source/tools/services/systemd.py]: Can't see a usage for this?

**Pull request merged**